lib/collection/src/collection_manager/holders/segment_holder.rs

```diff
@@ -434,17 +434,51 @@ impl<'s> SegmentHolder {
 
     /// Flushes all segments and returns maximum version to persist
     ///
+    /// Before flushing, this read-locks all segments. It prevents writes to all not-yet-flushed
+    /// segments during flushing. All locked segments are flushed and released one-by-one.
+    ///
     /// If there are unsaved changes after flush - detects lowest unsaved change version.
     /// If all changes are saved - returns max version.
     pub fn flush_all(&self, sync: bool) -> OperationResult<SeqNumberType> {
+        // Grab all locked segments to flush, in correct order
+        let segments = self
+            .segment_flush_ordering()
+            .map(|segment_id| self.segments.get(&segment_id).unwrap().get())
+            .collect::<Vec<_>>();
+
+        // Read-lock all segments at once before flushing, must prevent any writes to any segment
+        // That is to prevent any copy-on-write operation on two segments from occurring in between
+        // flushing the two segments. If that would happen, segments could end up in an
+        // inconsistent state on disk that is not recoverable after a crash.
+        //
+        // E.g.: we have a point on an immutable segment. If we use a set-payload operation, we do
+        // copy-on-write. The point from immutable segment A is deleted, the updated point is
+        // stored on appendable segment B.
+        // Because of flush ordering segment B (appendable) is flushed before segment A
+        // (not-appendable). If the copy-on-write operation happens in between, the point is
+        // deleted from A but the new point in B is not persisted. We cannot recover this by
+        // replaying the WAL in case of a crash because the point in A does not exist anymore,
+        // making copy-on-write impossible.
+        // Locking all segments prevents copy-on-write operations from occurring in between
+        // flushes.
+        let segment_reads = segments
+            .iter()
+            // Reverse so we lock from last to first
+            // Must lock non-appendable first to prevent deadlock with apply_points_to_appendable
+            .rev()
+            .map(|segment_lock| segment_lock.read())
+            .collect::<Vec<_>>();
```

Comment:
This currently locks non-appendable segments first, then appendable segments last. It is in reverse order versus our flushing order.

That is because the [`apply_points_to_appendable`](https://github.com/qdrant/qdrant/blob/eebfca01b6edda8f3ffc594f53e4c454a12332cb/lib/collection/src/collection_manager/holders/segment_holder.rs#L356-L403) function locks segments in this order: non-appendable first, then appendable.

We need to do this to prevent a deadlock.

Do we possibly want to implement a timeout/retry mechanism here to prevent surprise deadlocks? We already have [`aloha_random_write`](https://github.com/qdrant/qdrant/blob/eebfca01b6edda8f3ffc594f53e4c454a12332cb/lib/collection/src/collection_manager/holders/segment_holder.rs#L307-L309) but might want something similar to be on the safe side.

Reply:
Answer: yes, I'm implementing this now.

Reply:
Implemented in e870c70048dab940c1b9c7006427f55b88aba60b